lms/djangoapps/instructor/views/api.py

```diff
@@ -726,6 +729,377 @@ def extract_user_info(user):
+    results = data_access.make_total_query(clean_existing)
+    emails = [{"id": pair[0], "email":pair[1], "profileName":pair[2]} for pair in results]
+    if not make_csv:
+        response_payload = {
+            'course_id': course_id.to_deprecated_string(),
+            'data': emails,
+            'success': True,
+        }
+        return JsonResponse(response_payload)
+    else:
+        filename = time.strftime("%Y%m%d%H%M") + "emailSelection.csv"
+        email_pairs = [
+            [item['email'], item['profileName']]
+            for item in emails]
+        return instructor_analytics.csvs.create_csv_response(filename, ["email, name"], email_pairs)
+
+
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@require_level('instructor')
+@require_query_params(
+    filter="Type of filter",
+    entityName="Human readable name of entity"
+)
+def get_single_query(request, course_id, inclusion, query_type, state_type, state_id):
+    """
+    Makes and saves a single query
+    """
+    if inclusion is None or query_type is None or state_type is None or state_id is None:
+        return JsonResponse({
+            'success': False,
+        })
+
+    filtering = request.GET.get('filter')
+    entity_name = request.GET.get('entityName')
+    course_id = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    state_type_id = state_type + "/" + state_id
+    processed = _process_new_query(course_id, inclusion, query_type, state_type_id, filtering, entity_name)
+    if processed:
+        data_access.make_single_query.apply_async(args=(course_id, processed))
+        response_payload = {'course_id': course_id.to_deprecated_string(),
+                            'success': True,
+                            }
+        return JsonResponse(response_payload)
+    else:
+       # 500 on all other unexpected status codes.
+        log.error(
+            "Error parsing query-course id:{}, inclusion:{}, query_type:{}, state_type_id:{}, filtering:{},"
+            "entity_name:{}".format(
+                course_id, inclusion, query_type, state_type_id, filtering, entity_name
+            )
+        )
+        return HttpResponse(
+            "Error in parsing query.",
+            status=500,
+        )
+
+
+def _prune_course_tree(course_tree, include_pattern):
+    """
+    Helper functions for displaying problems and sections
+    """
+    new_tree = []
+    for child in course_tree:
+        if include_pattern.match(child['block_type']):
+            child['sub'] = _prune_course_tree(child['sub'], include_pattern)
+            new_tree.append(child)
+    return new_tree
+
+
+def _select_course_tree(course_tree, include_pattern):
+    """
+    Returns the 2 topmost parents of a node for displaying purposes
+    """
+    return _select_course_tree_recursive(course_tree, include_pattern, top_most=True)
+
+
+def _select_course_tree_recursive(course_tree,  # pylint: disable=dangerous-default-value
```

Comment:
I think we ought to figure out the return_container default value.  Currently this means that all invocations of `_select_course_tree_recursive` that don't supply return_container all share the _same_ list.